Add parsing for PE/COFF headers in libunwindstack

ronaldfw opened the pull request:
We add parsing of headers in PE/COFF files to our libunwindstack fork
in preparation for unwinding through PE/COFF frames. The parser is
skipping over some parts that we don't strictly need, such as several
members of the DOS header. Some other parts that we need are still
missing: We do not yet parse the section headers, which we will need to
extract the unwind information and other critical data. This wil be
added in a later change.

Tested: Unit tests.
Bug: http://b/192514457

beckerhe reviewed (approved):
Nice. Otherwise LGTM.

One more thing: It would be great to have a README file next to the test-DLLs explaining how they were built - just in case we need to rebuild them.

danielfenner commented on third_party/libunwindstack/PeCoffInterface.cpp:
```diff
@@ -0,0 +1,271 @@
+    return false;
+  }
+
+  coff_memory_.set_cur_offset(0x3c);
```
If this is something used frequently consider implementing `GetFully`.

ronaldfw replied:
Yes, you are correct. As far as I know, the initial offset will always be 0, so this will always be 0x3c, but it's for sure inconsistent with the rest. Let's keep it in generality with `offset` as the argument, so as you said this needs to be `offset + 0x3c`.

danielfenner commented on third_party/libunwindstack/include/unwindstack/PeCoffInterface.h:
```diff
@@ -0,0 +1,140 @@
+
+  ErrorData last_error_{ERROR_NONE, 0};
+
+  bool ParseDosHeader(uint64_t offset);
```
Not sure if this matters: Should this be 

bool ParseDosHeader(AddressType offset);

ronaldfw replied:
I don't think it matters. I haven't yet prototyped the 32-bit version, but I think AddressType only matters when we are actually dealing with stack memory in the target process, for example to get the return address etc. There it obviously matters whether we are just reading 4 bytes or incorrectly 8 bytes. This is the file offset and whether it's represented as uint32_t or uint64_t shouldn't matter.

danielfenner commented on third_party/libunwindstack/include/unwindstack/PeCoffInterface.h:
```diff
@@ -0,0 +1,140 @@
+
+template <typename AddressTypeArg>
+class PeCoffInterface {
+  static_assert(std::is_integral<AddressTypeArg>::value && !std::is_signed<AddressTypeArg>::value &&
```
This might be clearer:
` static_assert(std::is_same_v<AddressTypeArg, uint32_t> || std::is_same_v<AddressTypeArg, uint64_t>,"...")`

ronaldfw replied:
Thanks, good suggestion. Done.

ronaldfw commented on third_party/libunwindstack/include/unwindstack/PeCoffInterface.h:
```diff
@@ -0,0 +1,140 @@
+                "AddressTypeArg must be an unsigned integer and either 4-bytes or 8-bytes");
+
+ public:
+  PeCoffInterface(Memory* memory) : coff_memory_(memory) {}
```
Yes, done.

ronaldfw commented on third_party/libunwindstack/include/unwindstack/PeCoffInterface.h:
```diff
@@ -0,0 +1,140 @@
+  typedef AddressTypeArg AddressType;
+
+ protected:
+  PeCoffMemory coff_memory_;
+  DosHeader dos_header_;
+  CoffHeader coff_header_;
+  CoffOptionalHeader optional_header_;
+
+  ErrorData last_error_{ERROR_NONE, 0};
+
+  bool ParseDosHeader(uint64_t offset);
+  bool ParseNewHeader(uint64_t offset);
+  bool ParseCoffHeader(uint64_t offset);
+  bool ParseOptionalHeader(uint64_t offset);
+
+  bool ParseAllHeaders();
```
Yes, these should all be private for now. Not sure yet if we even need to derive from this class.

ronaldfw commented on third_party/libunwindstack/include/unwindstack/PeCoffInterface.h:
```diff
@@ -0,0 +1,140 @@
+  bool Init();
+  ErrorData LastError() const { return last_error_; }
+
+  typedef AddressTypeArg AddressType;
```
Done.

ronaldfw commented on third_party/libunwindstack/tests/PeCoffInterfaceTest.cpp:
```diff
@@ -0,0 +1,381 @@
+    EXPECT_GT(data.size(), 0u);
+    auto memory = std::make_unique<MemoryBuffer>();
+    EXPECT_TRUE(memory->Resize(data.size()));
+    memcpy(memory->GetPtr(0), data.data(), data.size());
```
Done.

ronaldfw commented on third_party/libunwindstack/tests/PeCoffInterfaceTest.cpp:
```diff
@@ -0,0 +1,381 @@
+  }
+
+  void InitPeCoffInterfaceFromFile() {
+    if (sizeof(typename PeCoffInterfaceType::AddressType) == 8) {
+      memory_ = ReadFile("libtest.dll");
+    } else if (sizeof(typename PeCoffInterfaceType::AddressType) == 4) {
+      memory_ = ReadFile("libtest32.dll");
+    }
```
Done, like the suggestion. Wasn't aware that "if constexpr" existed, that seems pretty useful. 

Also added this in other places where it makes sense.

ronaldfw commented:
Thanks for the reviews @beckerhe and @danielfenner ! I think I've addressed all comments, including adding a Makefile to compile the libtest.dll library.

ronaldfw commented:
I just noticed I'm missing a Makefile target for the 32-bit version of the library, will add ASAP.

ronaldfw commented:
> I just noticed I'm missing a Makefile target for the 32-bit version of the library, will add ASAP.

Ok, should be all complete now.
